### PR TITLE
Deduplicate various Active Record schema cache structures

### PR DESCRIPTION
**Pull request by casperisfine**
Superseeds: https://github.com/rails/rails/pull/35860 and https://github.com/rails/rails/pull/35855
Also https://github.com/rails/rails/pull/35875 is important for limiting the performance impact of the deduplication.

### Summary

Real world database schemas contain a lot of duplicated data.
Some column names like `id`, `created_at` etc can easily be repeated
hundreds of times. Same for SqlTypeMetada, most database will contain
only a limited number of possible combinations.

This result in a lot of wasted memory.

The idea here is to make these data sctructures immutable, use a registry
to substitute similar instances with pre-existing ones.

### Implementation details

This is mostly a refactor of https://github.com/rails/rails/pull/35860, instead of using the constructor arguments as cache key, I first instantiate the structure, and then look them up in the registry.

This makes for much cleaner code, and also provide a hook for deduplicating the internal values like https://github.com/rails/rails/pull/35860 did, which save a lot of CPU compared to deduplicating them on instantiation.

I also update the `SchemaCache` class to trigger the deduplication after the YAML schema has been loaded, this is because `YAML` use the `allocate.init_with(coder)` interface, hence bypassing the default constructor.

I'll add more comments on the diff.

### Efficiency

I posted various memory and CPU profiles in the related PRs. Today I shipped this PR as monkey patch on our app, and the resident memory difference is absolutely massive.

@rafaelfranca @kaspth @Edouard-chin @csfrancis

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/column.rb [hunk 1]
```diff
@@ -26,6 +28,17 @@ def initialize(name, default, sql_type_metadata = nil, null = true, table_name =
         @comment = comment
       end
 
+      def deduplicated
+        @name = -name
+        @table_name = -table_name if table_name
```
Comment (casperisfine) [inline activerecord/lib/active_record/connection_adapters/column.rb]: There are some test which instantiate a column without a `table_name`, I haven't been able to figure out if that's a realistic scenario or not. It seem to be related to dynamic columns when using `AR#select()`.

This makes me wonder if there's a risk of memory leak, some applications could possibly generate very dynamic queries and create new columns over and over.

File: activerecord/lib/active_record/connection_adapters/column.rb [hunk 2]
```diff
@@ -77,7 +90,7 @@ def hash
       protected
 
         def attributes_for_hash
-          [self.class, name, default, sql_type_metadata, null, table_name, default_function, collation]
+          [self.class, name, name.encoding, default, sql_type_metadata, null, table_name, default_function, collation]
```
Comment (casperisfine) [inline activerecord/lib/active_record/connection_adapters/column.rb]: This made a column encoding test fail. Turns out `String#hash` totally ignores the `String#encoding`.

```ruby
>> "foo".force_encoding('UTF-8').hash == "foo".force_encoding('EUC-JP').hash
=> true
```

Sounds a bit like an MRI bug to me.

File: activerecord/lib/active_record/connection_adapters/mysql/type_metadata.rb [hunk 3]
```diff
@@ -6,14 +6,22 @@ module MySQL
       class TypeMetadata < DelegateClass(SqlTypeMetadata) # :nodoc:
         undef to_yaml if method_defined?(:to_yaml)
 
+        include Deduplicable
+
         attr_reader :extra
 
-        def initialize(type_metadata, extra: "")
+        def initialize(type_metadata, extra: nil)
```
Comment (casperisfine) [inline activerecord/lib/active_record/connection_adapters/mysql/type_metadata.rb]: The vast majority of columns won't have an `extra`, `nil` seems a much better default value as we won't have to deduplicate an empty string.

File: activerecord/lib/active_record/connection_adapters/schema_cache.rb [hunk 4]
```diff
@@ -129,7 +129,16 @@ def marshal_dump
...
+        @indexes ||= {}
+
+        @columns = @columns.map { |table_name, columns| [-table_name, columns.map(&:deduplicate)] }.to_h
+        @columns_hash = @columns_hash.map do |table_name, columns|
+          columns_hash = columns.map { |column_name, column| [-column_name, column.deduplicate] }.to_h
+          [-table_name, columns_hash]
+        end.to_h
+        @primary_keys = @primary_keys.map { |table_name, column_name| [-table_name, -column_name] }.to_h
+        @data_sources = @data_sources.transform_keys { |table_name| -table_name }
+        @indexes = @indexes.map { |table_name, indexes| [-table_name, indexes] }.to_h
```
Comment (casperisfine) [inline activerecord/lib/active_record/connection_adapters/schema_cache.rb]: I think we should make `IndexDefinition` Deduplicable as well, but it doesn't have a `==` nor `hash` method, so I think it would be better handled in a followup.

File: activerecord/test/active_record/connection_adapters/fake_adapter.rb [hunk 5]
```diff
@@ -32,7 +32,9 @@ def merge_column(table_name, name, sql_type = nil, options = {})
           name.to_s,
           options[:default],
           fetch_type_metadata(sql_type),
-          options[:null])
+          options[:null],
+          table_name,
```
Comment (casperisfine) [inline activerecord/test/active_record/connection_adapters/fake_adapter.rb]: This test wan't passing the table name, resulting in a quite unrealistic scenario (IMHO)

File: activerecord/test/cases/base_test.rb [hunk 6]
```diff
@@ -1131,11 +1131,14 @@ def test_benchmark_with_use_silence
...
     # preheat cache
     c1 = Post.connection.schema_cache.columns("posts")
+    refute_equal 0, Post.connection.schema_cache.size
+
     ActiveRecord::Base.clear_cache!
+    assert_equal 0, Post.connection.schema_cache.size
+
     c2 = Post.connection.schema_cache.columns("posts")
-    c1.each_with_index do |v, i|
-      assert_not_same v, c2[i]
```
Comment (casperisfine) [inline activerecord/test/cases/base_test.rb]: The interning made this way of testing the schema was reloaded fail. 

I refactored the test, I believe it still offer the same confidence.

File: activerecord/test/models/contact.rb [hunk 7]
```diff
@@ -10,14 +10,14 @@ def self.extended(base)
...
-      column :preferences,    :string
-      column :alternative_id, :integer
+      column :id,             'integer'
+      column :name,           'string'
+      column :age,            'integer'
+      column :avatar,         'binary'
+      column :created_at,     'datetime'
+      column :awesome,        'boolean'
+      column :preferences,    'string'
+      column :alternative_id, 'integer'
```
Comment (casperisfine) [inline activerecord/test/models/contact.rb]: I might be wrong but I believe `sql_type` is never a symbol in real world scenario, so I took the liberty to refactor this test data.

**Comment by casperisfine** [comment]
Rebased to fix conflict with https://github.com/rails/rails/pull/35890

Now that `Column` instance no longer contain the table name, the deduplication should be even more efficient.

**Comment by casperisfine** [comment]
So with  #35890 profiles show that retained `ActiveRecord::ConnectionAdapters::MySQL::Column` instances have been further reduced from `7768 (807.89 kB)` to `2550 (265.22 kB)`. Not too bad!

File: activerecord/lib/active_record/connection_adapters/schema_cache.rb (hunk 4 above)
Comment (kaspth) [reply activerecord/lib/active_record/connection_adapters/schema_cache.rb]: Agreed on `IndexDefinition`!

I find this code quite hard to parse. The meaning is hidden behind the processing and the individual `-` are lost — plus the differences for when to use `deduplicate`. Would something like this be clearer?

```ruby
  @columns      = deep_deduplicate(@columns) { |columns| columns.map(&:deduplicate) }
  @columns_hash = deep_deduplicate(@columns_hash) { |columns| deep_deduplicate(columns, &:deduplicate) }
  @primary_keys = deep_deduplicate(@primary_keys)
  @data_sources = deep_deduplicate(@data_sources, &:itself)
  @indexes      = deep_deduplicate(@indexes, &:itself)
end

def deep_deduplicate(values, &block)
  block ||= :-@.to_proc
  values.transform_keys(&:-@).transform_values(&block)
end
```

The `:-@.to_proc` should probably be cleaned up a bit somehow. Also not sure how wasteful it is to loop through the probably gargantuan hashes twice.

Another thing we could do is make the deduplicateable objects respond to `-@` instead of `deduplicate`. Then we don't have to make the difference between them and follow Ruby's new semantics.

File: activerecord/lib/active_record/connection_adapters/deduplicable.rb [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+          @registry ||= {}
+        end
+
+        def new(*)
+          super.deduplicate
+        end
+      end
+
+      def deduplicate
+        self.class.registry[self] ||= self.deduplicated
```
Comment (kaspth) [inline activerecord/lib/active_record/connection_adapters/deduplicable.rb]: `self.deduplicated` can just be `deduplicated`, no? Also don't see why `deduplicated` should be a public method.

File: activerecord/lib/active_record/connection_adapters/schema_cache.rb (hunk 4 above)
Comment (casperisfine) [reply activerecord/lib/active_record/connection_adapters/schema_cache.rb]: > Another thing we could do is make the deduplicateable objects respond to `-@` instead of `deduplicate`. Then we don't have to make the difference between them and follow Ruby's new semantics.

That's what I wanted to do initially, however I though that since numeric types respond to that same interface for a very different meaning made me reconsider it. So I'm split on this.

File: activerecord/lib/active_record/connection_adapters/deduplicable.rb (hunk 8 above)
Comment (casperisfine) [reply activerecord/lib/active_record/connection_adapters/deduplicable.rb]: Yeah good point, I missed that.

File: activerecord/lib/active_record/connection_adapters/schema_cache.rb (hunk 4 above)
Comment (casperisfine) [reply activerecord/lib/active_record/connection_adapters/schema_cache.rb]: > Agreed on IndexDefinition!

Actually I wonder about these. Turns out our internal SchemaCache implementation doesn't persist them, our prod workers have 0 of them in memory. So I wonder how useful they actually are.

**Comment by casperisfine** [comment]
I rebased following @kamipo 's refactor in https://github.com/rails/rails/commit/59d32a6eab144b808bdd9d96b6e6c2dc32e66f32 and also based on @kaspth suggestions.

File: activerecord/lib/active_record/connection_adapters/schema_cache.rb [hunk 9]
```diff
@@ -129,10 +129,29 @@ def marshal_dump
 
       def marshal_load(array)
         @version, @columns, @columns_hash, @primary_keys, @data_sources, @indexes, @database_version = array
-        @indexes = @indexes || {}
+        @indexes ||= {}
+
+        @columns = deep_deduplicate(@columns)
+        @columns_hash = deep_deduplicate(@columns_hash)
```
Comment (casperisfine) [inline activerecord/lib/active_record/connection_adapters/schema_cache.rb]: @kaspth another thing: I think we should simply stop storing `columns_hash` in the schema_cache dump, it's very easily rebuilt from `columns`. 

That would make parsing much faster, and would save lots of allocations.

Comment (kaspth) [reply activerecord/lib/active_record/connection_adapters/schema_cache.rb]: I’d definitely be game to see a PR for that!

File: activerecord/lib/active_record/connection_adapters/schema_cache.rb (hunk 4 above)
Comment (kaspth) [reply activerecord/lib/active_record/connection_adapters/schema_cache.rb]: iirc, indexes were added to the cache in Rails 6.

**Comment by kaspth** [comment]
Love it!
